Microsoft.Quantum.Canon/Utils/Multiplexor.qs

```diff
@@ -0,0 +1,285 @@
+    ///
+    /// # Remarks
+    /// `coefficients` will be padded with elements $\theta_j = 0.0$ if 
+    /// fewer than $2^n$ are specified.
+    ///
+    /// # References
+    /// - Synthesis of Quantum Logic Circuits
+    ///   Vivek V. Shende, Stephen S. Bullock, Igor L. Markov
+    ///   https://arxiv.org/abs/quant-ph/0406176
+    operation MultiplexorZ(coefficients: Double[], control: BigEndian, target: Qubit) : () {
+        body{
+            // pad coefficients length to a power of 2.
+            let maxCoefficients = 2^(Length(control));
+            let padZeros = new Double[maxCoefficients - Length(coefficients)];
+            
+            if(maxCoefficients == 1){
+                // Termination case
+                Exp([PauliZ], coefficients[0], [target]);
+            }
+            else{
+                // Compute new coefficients.
+                let (coefficients0, coefficients1) = MultiplexorZComputeCoefficients_(coefficients + padZeros);
+                MultiplexorZ(coefficients0, BigEndian(control[1..Length(control)-1]), target);
+                CNOT(control[0], target);
+                MultiplexorZ(coefficients1, BigEndian(control[1..Length(control)-1]), target);
+                CNOT(control[0], target);
+            }
+        }
+        adjoint auto
+        controlled (controlRegister) {
+            // pad coefficients length to a power of 2.
+            let maxCoefficients = 2^(Length(control));
+            let padZeros = new Double[maxCoefficients - Length(coefficients)];
+
+            let controlZeros = new Double[maxCoefficients];
+            let (coefficients0, coefficients1) = MultiplexorZComputeCoefficients_(controlZeros+coefficients+padZeros);
+            
+            MultiplexorZ(coefficients0, control, target);
+            (Controlled X)(controlRegister, target);
+            MultiplexorZ(coefficients1, control, target);
+            (Controlled X)(controlRegister, target);
+        }
+        adjoint controlled auto
+    }
+
+    /// # Summary
+    /// Diagonal unitary operation $U$ that applies a complex phase 
+    /// $e^{i \theta_j}$ on the $n$-qubit number state $\ket{j}$.
+    ///
+    /// $U = \sum^{2^n-1}_{j=0}e^{i\theta_j}\ket{j}\bra{j}$.
+    ///
+    /// # Input
+    /// ## coefficients
+    /// Array of up to $2^n$ coefficients $\theta_j$. The $j$th coefficient 
+    /// indexes the number state $\ket{j}$ encoded in big-endian format. 
+    ///
+    /// ## control
+    /// $n$-qubit control register that encodes number states $\ket{j}$ in
+    /// big-endian format.
+    ///
+    /// # Remarks
+    /// `coefficients` will be padded with elements $\theta_j = 0.0$ if 
+    /// fewer than $2^n$ are specified.
+    ///
+    /// # References
+    /// - Synthesis of Quantum Logic Circuits
+    ///   Vivek V. Shende, Stephen S. Bullock, Igor L. Markov
+    ///   https://arxiv.org/abs/quant-ph/0406176
+    operation DiagonalUnitary(coefficients: Double[], qubits: BigEndian) : () {
```

Comment:
I'd suggest making this `ApplyDiagonalUnitary`, such that `DiagonalUnitary : Double[] -> (BigEndian => () : Adjoint, Controlled)`.

Reply:
Renamed